File: Ing. en Informatica/2° Año/2º Semestre/Metodos Numericos/Codigos/En C/8_14/Metodos Abiertos/Secante.c
#include <stdio.h>
#include <math.h>
#define MAX_ITER 1000000

// Declaraciones de funciones
double fun_f(double x);
double secante(double x0, double x1, double epsilon, double *error, int *iter, int presicion);

int main() {
    double x0=0, x1=0, root=0, epsilon=0, error=0;
    int iter=0;

    // Solicitar al usuario los valores iniciales y la tolerancia
    printf("Ingrese el valor inicial x0: ");
    scanf("%lf", &x0);
    printf("Ingrese el valor inicial x1: ");
    scanf("%lf", &x1);
    printf("Ingrese la tolerancia (epsilon): ");
    scanf("%lf", &epsilon);


    // Calcular la cantidad de decimales según la tolerancia
    int precision = -log10(epsilon) + 1;

    // Llamar a la función secant_method para encontrar la raíz
    root = secante(x0, x1, epsilon, &error, &iter, precision);

    // Mostrar el resultado
    printf("\nEl punto fijo Xn=%.*lf, con un error de=%.*lf\nLa cantidad de Iteraciones fueron: %d.\n", precision,
           root, precision, error, iter);
    printf("\n%.*lf %c %.*lf\n", precision, root, 241, precision, error);

    return 0;
}

// Implementación de la función secant_method
double secante(double x0, double x1, double epsilon, double *error, int *iter, int presicion){
    double x2=0;

    *iter = 0; // Inicializar el contador de iteraciones

    // Realizar iteraciones hasta que se cumple el criterio de parada
    do {
        x2 = x1 - (fun_f(x1) * (x1 - x0)) / (fun_f(x1) - fun_f(x0)); // Calcular el siguiente punto
        *error = fabs(x2 - x1); // Calcular el error absoluto
        x0 = x1; // Actualizar x0 con x1
        x1 = x2; // Actualizar x1 con x2
        (*iter)++; // Incrementar el contador de iteraciones

        // Imprimir información sobre la iteración (comentada para desactivar)
        // printf("Iteración %d: x = %.*lf, error = %.*lf\n", *iter, presicion, x1, presicion, *error);

    } while ((*error) > epsilon && (*iter) < MAX_ITER); // Condición de parada

    if (*iter == MAX_ITER)
        printf("Hubo una salida por el máximo de iteraciones establecido.\n");

    return x2; // Devolver la raíz aproximada
}

double fun_F(double d){//n se define primero la funcion y luego se puede ejecutar el programa.
    //return log(d) + exp( sin( d ) ) - d ;//Modelo para probar los metodos
    //return 4-pow(d,2) ;//Ejercicio 1 Guia 2
    return pow (d,5)- 3 * pow(d, 3) - 2 * pow(d, 2) + 2;//Ejercicio 2)a Guia 2
    //  Las 2 siguientes son constante, por eso quedan igual.
    //return cos(sin(d));//Ejercicio 2)b Guia 2
    //return pow(d,d-cos(d));//Ejercicio 2)c Guia 2
}
double fun_G(double d){//n se define primero la funcion y luego se puede ejecutar el programa.
    //return log(d) + exp( sin( d ) ) ;//Modelo para probar los metodos
    //return (-(sin(d)+1));//Funcion random del Agus.
    //return 4-pow(d,2) ;//Ejercicio 1 Guia 2
    return pow( (3 * pow(d, 3) + 2 * pow(d, 2) - 2), 1.0/5);//Ejercicio 2)a Guia 2
    //  Las 2 siguientes son constante, por eso quedan igual.
    //return cos(sin(d));//Ejercicio 2)b Guia 2
    //return pow(d,d-cos(d));//Ejercicio 2)c Guia 2
}

File: Ing. en Informatica/2° Año/2º Semestre/Metodos Numericos/Codigos/En Java/utils/abiertos/Secante.java
package utils.abiertos;

import java.text.DecimalFormat;
import java.util.Scanner;

public class Secante {

    private double x0=0;
    private double xv=0;
    private double xn=0;
    private double error=0;
    private int cont=0;
    private double tolerancia=0;
    private final int MAX_ITER = 1000000;
    public Secante(double x0, double xv, double tol) {
        this.x0=x0;
        this.xv=xv;
        this.tolerancia=tol;
    }
    public void calcularRaiz(){
        cont=0;
        do
        {
            xn = xv - (fun_F(xv) * (xv - x0)) / (fun_F(xv) - fun_F(x0)); // Calcular el siguiente punto
            error = Math.abs(xn - xv); // Calcular el error
            x0 = xv; // Actualizar x0 con xv
            xv = xn; // Actualizar xv con xn
            cont++; // Incrementar el contador de iteraciones
            //System.out.println("itersacion="+cont);
        } while ( (error > tolerancia) && (cont < (MAX_ITER/5)) ) ;

        if(cont==MAX_ITER)
            System.out.println("Hubo una salida por el maximo de salidas establecido.\n");

        DecimalFormat df = new DecimalFormat("0.000000000000000"); // Define el formato con 10 decimales
        String formattedError = df.format(getError());// Calcular la cantidad de decimales según la tolerancia
        System.out.println("El punto fijo mediante Secante Xn="+getXn()+",con un de error="+getError()+"\nLa cantidad de Iteraciones fueron: "+getCont()+".");
        System.out.printf(getXn()+ "+-" + formattedError);

    }
    public double getError() {return error;}
    public int getCont() {return cont;}
    public double getX0() {return x0;}

    public double getXn() {return xn;}

    public double getXv() {return xv;}

    double fun_F(double d){//n se define primero la funcion y luego se puede ejecutar el programa.
        //return Math.log(d) + Math.exp( Math.sin( d ) ) - d ;//Modelo para probar los metodos
        //return 4-Math.pow(d,2) ;//Ejercicio 1 Guia 2
        //return Math.pow (d,5)- 3 * Math.pow(d, 3) - 2 * Math.pow(d, 2) + 2;//Ejercicio 2)a Guia 2
        //  Las 2 siguientes son constante, por eso quedan igual.
        //return Math.cos(Math.sin(d));//Ejercicio 2)b Guia 2
        //return Math.pow(d,d-Math.cos(d));//Ejercicio 2)c Guia 2


        return Math.cos(3+d+4*d);
    }
    double fun_G(double d){//n se define primero la funcion y luego se puede ejecutar el programa.
        //return Math.log(d) + Math.exp( Math.sin( d ) ) ;//Modelo para probar los metodos
        //return (-(Math.sin(d)+1));//Funcion random del Agus.
        //return 4-Math.pow(d,2) ;//Ejercicio 1 Guia 2
        //return Math.pow( (3 * Math.pow(d, 3) + 2 * Math.pow(d, 2) - 2), 1.0/5);//Ejercicio 2)a Guia 2
        //  Las 2 siguientes son constante, por eso quedan igual.
        //return Math.cos(Math.sin(d));//Ejercicio 2)b Guia 2
        //return Math.pow(d,d-Math.cos(d));//Ejercicio 2)c Guia 2


        //PRUEBA:
        double h = 0.000001;
        return ((fun_F(d + h) - fun_F(d)) / h);
    }
}
